### PR TITLE
Add hook for pyopencl and set default spec file encoding (fixes python 3.5 issue)

### PR DESCRIPTION
**Pull request by janbraiins**
- this adds necessary include paths that are part of the pyopencl
  metadata see (pyopencl/__init__.py::_find_pyopencl_include_path())

NOTE: it works with the latest pyopencl that has a typo fixed (merged just today: https://github.com/pyopencl/pyopencl/pull/157)

Thank you for including this.

Best regards,

Jan

### CODE REVIEW

**Comment by honzik666** [comment]
I have implemented what has been discussed in the past 2 comments. Let me know whether this is acceptable. I have tested it on 2.7, 3.4 and 3.5 with a spec file containing some UTF-8 characters.

File: PyInstaller/hooks/hook-pyopencl.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+# Distributed under the terms of the GNU General Public License with exception
+# for distributing bootloader.
+#
+# The full license is in the file COPYING.txt, distributed with this software.
+#-----------------------------------------------------------------------------
+
+# Hook for the pyopencl module: https://github.com/pyopencl/pyopencl
+
+from PyInstaller.utils.hooks import copy_metadata
+datas = copy_metadata('pyopencl')
```
Comment (htgoebel) [inline PyInstaller/hooks/hook-pyopencl.py]: #2305 also collects data files. What about these?

File: PyInstaller/compat.py [hunk 2]
```diff
@@ -889,3 +891,30 @@ def check_requirements():
             raise SystemExit('PyInstaller cannot check for assembly dependencies.\n'
                              'Please install PyWin32.\n\n'
                              'pip install pypiwin32\n')
+
+
+@contextmanager
+def open_pysrc(filename):
```
Comment (htgoebel) [inline PyInstaller/compat.py]: This is a good idea. I suggest moving this into modulgraph, where is similar code which could be updated.

File: PyInstaller/compat.py [hunk 3]
```diff
@@ -889,3 +891,30 @@ def check_requirements():
...
+    if getattr(tokenize, 'open', None) is not None:
+        print('USING tokenize')
```
Comment (htgoebel) [inline PyInstaller/compat.py]: Please remove this line. I you feel this is really necessary, it should be a log message of level DEBUG and be much more spelling.

File: PyInstaller/compat.py [hunk 4]
```diff
@@ -889,3 +891,30 @@ def check_requirements():
...
+        pysrc_f = tokenize.open(filename)
+    else:
+        print('Guessing encoding')
```
Comment (htgoebel) [inline PyInstaller/compat.py]: same here

File: PyInstaller/compat.py [hunk 5]
```diff
@@ -889,3 +891,30 @@ def check_requirements():
...
+        from PyInstaller.lib.modulegraph.util import guess_encoding
```
Comment (htgoebel) [inline PyInstaller/compat.py]: In Pyinstaller all imports belong to the top of the file.

File: PyInstaller/hooks/hook-pyopencl.py (hunk 1 above)
Comment (honzik666) [inline PyInstaller/hooks/hook-pyopencl.py]: Now I am a bit confused, #2305 has been closed as a duplicate of this PR. I have come across #2305 randomly while looking for some other pyinstaller related issues.

Comment (htgoebel) [inline PyInstaller/hooks/hook-pyopencl.py]: Yes, I closed #2305. But that one doffers from yout PR, this is why I'm asking you about collecting the data-files which is done there.

File: PyInstaller/compat.py (hunk 5 above)
Comment (honzik666) [inline PyInstaller/compat.py]: ok, I just saw imports occassionally interleaved throughout the source, so I was thinking it's not a problem to use it this way. I will move it to the top.

File: PyInstaller/compat.py (hunk 3 above)
Comment (honzik666) [inline PyInstaller/compat.py]: This is my bad, I will change it to debug log message.

File: PyInstaller/compat.py (hunk 2 above)
Comment (honzik666) [inline PyInstaller/compat.py]: ok, will do. The read_pysrc() method will stay in compat.py

File: PyInstaller/hooks/hook-pyopencl.py (hunk 1 above)
Comment (honzik666) [inline PyInstaller/hooks/hook-pyopencl.py]: Sorry, I didn't notice. Yes, the data files are to be collected, too. I will update the changeset accordingly, too

**Comment by htgoebel** [comment]
I cherry-picked the first commit.
For the second (unrelated) commit I added a test-case (c6a7776a4566f14b48795123bfd6c3e6ee19149a) which passes with the current code. So I can not see the problem, the second commit tries to fix. Can you please verify this?

**Comment by htgoebel** [comment]
@honzik666 pinging. What about this pull-request?

**Comment by Legorooj** [comment]
Closing due to author inactivity

**Pull request closed without merging**